### PR TITLE
Examples: Add ground projected env map

### LINKED ISSUE #23512: Ground projected HDR environment map
**Is your feature request related to a problem? Please describe.**

It would be nice to have HDR environment maps mapped onto flat planes or circles that are flatted near the bottom centre.

**Describe the solution you'd like**

It should be possible to set a vertical ground height and also a bias parameter that determines how much the sphere is flatted to this ground plane.

Demo:
https://twitter.com/repalash/status/1494270181027508224

**Describe alternatives you've considered**

Modelling a flatted sphere, but that is a lot of work.

### PR DESCRIPTION
**Pull request by FarazzShaikh**
Fixed #23512

**Description**

Adds ground projected Env map helper to `examples/jsm/objects/GroundProjectedEnv.js`. Impl based on `@react-three/drei`'s. Samples an env map based on ray casting result from camera to a SDF sphere (for background) and disc (for the ground).

Example use:
```js
import { GroundProjectedEnv } from "three/jsm/objects/GroundProjectedEnv";
...
const env = new GroundProjectedEnv(envMap);
env.scale.setScalar(100);
scene.add(env);
...
env.radius = ...;
env.height = ...;
```
Example added to `examples/webgl_materials_envmaps_ground-projected.html`
image

### CODE REVIEW

**Review by gkjohnson — COMMENTED** [review]
This looks great! Here's a live link to the demo:

http://raw.githack.com/FarazzShaikh/three.js/feat/envmaps-ground-projected/examples/index.html?q=ground#webgl_materials_envmaps_ground-projected

Two nits:

- Can we remove the build files and package-lock changes from the PR?
- It would be nice to use the project code style for the js and html files but as far as I understand that is not blocking for examples.

**Comment by FarazzShaikh** [comment]
Thanks Garrett, Of course I can make those changes!

File: examples/webgl_materials_envmaps_ground-projected.html [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		<script type="module">
+			import * as THREE from "three";
+
+			import Stats from "./jsm/libs/stats.module.js";
+
+			import { GUI } from "./jsm/libs/lil-gui.module.min.js";
+			import { OrbitControls } from "./jsm/controls/OrbitControls.js";
+			import { HDRCubeTextureLoader } from "./jsm/loaders/HDRCubeTextureLoader.js";
+			import { RGBMLoader } from "./jsm/loaders/RGBMLoader.js";
+			import { DebugEnvironment } from "./jsm/environments/DebugEnvironment.js";
```
Comment (Mugen87) [inline examples/webgl_materials_envmaps_ground-projected.html:46]: It seems the imports of `HDRCubeTextureLoader`, `RGBMLoader` and `DebugEnvironment` can be removed.

File: examples/jsm/misc/GroundProjectedEnv.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import * as THREE from "three";
```
Comment (Mugen87) [inline examples/jsm/misc/GroundProjectedEnv.js:1]: In the example modules, we normally only import what's required. For this particular use case, it should be `Mesh`,  `IcosahedronGeometry`, `ShaderMaterial` and `DoubleSide`.

**Review by Mugen87 — APPROVED** [review]

**Comment by LeviPesin** [comment]
Please fix broken lint.

File: examples/jsm/misc/GroundProjectedEnv.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
+import {
+	Texture,
```
Comment (LeviPesin) [inline examples/jsm/misc/GroundProjectedEnv.js:2]: `Texture` import is unused.

Comment (FarazzShaikh) [reply examples/jsm/misc/GroundProjectedEnv.js:2]: Was used in JSDoc comments, removed the comments and the unused import

**Comment by FarazzShaikh** [comment]
ran `npm run lint-fix` on `examples/jsm/misc/GroundProjectedEnv.js` no lint problems were reported by the lint command on other files

**Comment by LeviPesin** [comment]
> no lint problems were reported by the lint command on other related files

There are problems remaining in `files.json` (just revert the changes to it and add example), GLSL code in the class (like missing spaces, you can fix that manually), and in the example (also just missing spaces).

**Comment by FarazzShaikh** [comment]
Using car instead of the helmet now. Limited the controls and GUI sliders. Fixed GLSL formatting in the class

image

File: examples/files.json [hunk 4]
```diff
@@ -141,6 +141,7 @@
 		"webgl_materials_displacementmap",
 		"webgl_materials_envmaps",
 		"webgl_materials_envmaps_exr",
+		"webgl_materials_envmaps_ground-projected",
```
Comment (LeviPesin) [inline examples/files.json:144]: I suggest to replace `-` in the example name with `_`.

**Comment by LeviPesin** [comment]
Lint is again broken in the class and the example.

File: examples/files.json (hunk 4 above)
Comment (FarazzShaikh) [reply examples/files.json:144]: I believe that would make the sidebar section title `materials / envmaps / ground / projected` and not `materials / envmaps / ground-projected`

**Comment by WestLangley** [comment]
Screen Shot 2022-07-06 at 1 06 41 PM

Set limits to the controls to prevent that from happening.
Probably prohibit panning.
Also, further restrict the polar angle to 80 degrees or less.

**Comment by LeviPesin** [comment]
> Ran lint fix command on the class, do you have a better way to lint GLSL? It looks good to me

The same rules of linting JS apply to GLSL (spaces between parentheses and arguments and empty lines padding in blocks).

**Comment by FarazzShaikh** [comment]
i think that should do it?

**Comment by FarazzShaikh** [comment]
Sure i can add in a rotation option. 

I will also move the class to `examples/jsm/objects` since it is quite similar to the `Sky` class and thats where that is.

**Comment by WestLangley** [comment]
@FarazzShaikh Thank your for your PR. It is not following three.js best practices, however.

@Mugen87

1. TBH, I do not think we need another LDR cube map in the examples. An existing three.js HDR environment map should be used, instead.

2. The example should also be properly tone mapped, with exposure explicitly set.

3. Why it is using and AO map as a shadow with `MultiplyBlending` and no tone mapping? This is not standard three.js workflow.

**Comment by FarazzShaikh** [comment]
@WestLangley I have added tone mapping and exposure settings to the example. Is there a best practices doc somewhere? I might have missed it

The effect requires a large flat env to look convincing, the `Park2` cube map works fine but has a lot of shadows that break the illusion and make it look CGI/fake. i would be happy to remove the env map and use one thats already in the repo.

**Comment by FarazzShaikh** [comment]
No worries, will revert rotation feature. My bad i forgot the reflections wouldn't rotate even if the env did. Perhaps we can add it back later when env map rotation is supported in core.

**Review by gkjohnson — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Comment by blokingsmall** [comment]
image

In my project

`     const HDR = useLoader(THREE.TextureLoader, './bg_1.jpeg')
	HDR.encoding = THREE.sRGBEncoding
	var env = new GroundProjectedEnv( HDR );
	env.scale.setScalar( 100 );
	three.scene.add( env )
`

**Comment by Mugen87** [comment]
@blokingsmall Please use the [forum](https://discourse.threejs.org/) first. If it turns out to be an issue with the implementation, this should be reported as a new GitHub issue.

**Comment by blokingsmall** [comment]
@Mugen87 ok i'll try it

**Comment by Mugen87** [comment]
Forum topic: https://discourse.threejs.org/t/groundprojectedenv-seems-to-be-some-problems/41625

**Comment by WestLangley** [comment]
/ping @FarazzShaikh See #24520. I am assuming the issue needs further investigation.

**Comment by gkjohnson** [comment]
According to [this comment](https://discourse.threejs.org/t/groundprojectedenv-seems-to-be-some-problems/41625/10) disabling mipmaps on the env map fixes the issue. It seems react-three-fiber automatically enables mipmaps on textures. But the issue here is that the UVs for sampling are dynamically generated and the derivative of the passed value is used to derive which mipmap to use. And because it's such a stark jump from UV 0.0 to 1.0 at the wrap around seam. This leads to a very high dervaitive value and likely the lowest LoD mipmap being selected. Normally this isn't an issue when mipmaps are disabled or the UV seam is at a triangle boundary. See here:

| Wireframe UVs | Solid UVs |
|---|---|
| image | image |

You can see that the UV seam occurs mid triangle.

And if you take a look at the derivatives per-pixel you can see the seam in the same spot:



My recommendation here is to call this "not a bug" and require users to pass in a texture with mipmaps disabled or generate a geometry such that this UV seam is always on a triangle edge.

Also if you're a user submitting a bug report please make sure it's using vanilla three.js. Giving repros based on environments like react-three-fiber that adjust defaults values especially when it's not clear what's being used... it makes understanding issues like this so much harder and is why we couldn't repro it in the official example... 😭

**Comment by WestLangley** [comment]
>or generate a geometry such that this UV seam is always on a triangle edge.

@gkjohnson Are you implying that replacing `IcosahedronGeometry` with `SphereGeometry( 1, 32, 16 )` will fix the problem?

**Comment by gkjohnson** [comment]
> Are you implying that replacing IcosahedronGeometry with SphereGeometry( 1, 32, 16 ) will fix the problem?

I'm not so familiar with the internals of the grounded env map but I don't believe it's so simple - at least not without some more extensive changes to the the class. Just from the wireframe of the mesh it looks like it serves as more of just a rasterization surface with procedural computations being used for the envmap "ground" projection. If the surface of the mesh is used directly then it's likely you'll see artifacts as the result of the geometry resolution in the visuals of the env map though it might not be all that noticeable. I'll have to let @FarazzShaikh speak on what might be doable here - I've just run into this issue of unexpected procedural mipmap calculations before 😁

**Comment by WestLangley** [comment]
>or generate a geometry such that this UV seam is always on a triangle edge.

Thing is, the shader does not use the geometry uv's. So I am not sure the alignment of the triangles matters...

**Comment by gkjohnson** [comment]
It doesn't currently but it could if the geometry were formed appropriately.

**Comment by WestLangley** [comment]
>It doesn't currently but it could if the geometry were formed appropriately.

I see. So to rephrase your proposed solution:

1.  modify the shader to use the geometry UVs, and
2. insure the new sphere geometry triangles are aligned with the seam.

If you are correct, and that works, then we may need to revisit the built-in "cube shader" on which this is based. I do not recall seeing the cube shader have the same artifacts, however...

**Comment by gkjohnson** [comment]
Yes both of those are required to support the solution of "generate a geometry such that this UV seam is always on a triangle edge".

> we may need to revisit the built-in "cube shader" on which this is based. I do not recall seeing the cube shader have the same artifacts, however...

I don't know all the shader code that's involved in the existing implementations but from a brief check it looks like the [cube map sampling](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/envmap_fragment.glsl.js#L39) uses the builtin [textureCube](https://thebookofshaders.com/glossary/?search=textureCube) function so it shouldn't be an issue. Though I don't about the `ENVMAP_TYPE_CUBE_UV` case.

**Comment by drcmda** [comment]
> Giving repros based on environments like react-three-fiber that adjust defaults values especially when it's not clear what's being used

hmm, that would be very odd. we do not change defaults, a texture in fiber is just a THREE.Texture. fiber doesn't know what three is technically, i don't see how it would decide to change mipmaps of all things.

**Comment by FarazzShaikh** [comment]
I will take a detailed look at this when some time opens up between work. Admittedly I do not have very deep understanding of low level Textures and Mipmaps

**Comment by WestLangley** [comment]
@gkjohnson I misspoke, I meant to say `equirect shader` -- not `cube shader`.

**Comment by WestLangley** [comment]
@drcmda I am not sure the PMREM-related code in `GroundProjectedEnv.js` belongs there...

**Comment by gkjohnson** [comment]
> hmm, that would be very odd. we do not change defaults, a texture in fiber is just a THREE.Texture. it doesn't know what three is technically, i don't see how it would decide to change mipmaps of all things.

My mistake. It looks like the difference here is that the official example uses RGBELoader whereas the repro case is using TextureLoader, which is likely where the difference in mipmap setting comes from.

> @gkjohnson I misspoke, I meant to say equirect shader -- not cube shader.

As far as I understand the equirect map is converted to a cube map:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLBackground.js#L29

**Comment by WestLangley** [comment]
>As far as I understand the equirect map is converted to a cube map:

Not in this example.

The problem is caused by `uv = equirectUv( direction )` when mips are in use. And as you said, the problem occurs at the UV seam.

**Pull request merged**